prac_07/project_management.py

```diff
@@ -0,0 +1,124 @@
+            for i, project in enumerate(projects):
+                print(i, project)
+            project_choice = int(input("Project choice: "))  # error checking needed
+            print(projects[project_choice])
+            update_project(projects, project_choice)
+        else:
+            print("Invalid choice")
+        print(MENU)
+        choice = input(">>> ").upper()
+    print("Thank you for using custom-built project management software.")
+    save_projects(FILENAME, projects)
+
+
+def update_project(projects, project_choice):
+    new_percentage = input("New percentage: ")
+    new_priority = input("New priority: ")
+    if new_percentage != "":  # can extract function + can change if statement "" = false
+        try:
+            projects[project_choice].completion = int(new_percentage)
+        except ValueError:
+            print("Percentage must be a number")
+    if new_priority != "":
+        try:
+            projects[project_choice].priority = int(new_priority)
+        except ValueError:
+            print("Priority must be  number")
+
+
+def get_new_project_details():
+    name = get_valid_string("Name: ")
+    date_string = input("Start date (dd/mm/yyyy): ")  # error checking needed
+    start_date = format_date(date_string)
+    priority = get_valid_number(int, "Priority: ")
+    cost = get_valid_number(float, "Cost estimate: $")
+    completion = get_valid_number(int, "Percent complete: ")
+    return name, start_date, priority, cost, completion
+
+
+def get_valid_string(prompt):
+    string = input(prompt)
+    while string == "":
+        print("Invalid string cannot be empty.")
+        string = input(prompt)
+    return string
+
+
+def get_valid_number(number_type, prompt):
+    is_valid = False
+    while not is_valid:
+        try:
+            number = number_type(input(prompt))
+            if number < 0:
+                print("Invalid number should be >= 0.")
+            else:
+                is_valid = True
+        except ValueError:
+            print("Invalid number")
+    return number
+
+
+def format_date(date_string):
+    formatted_date = datetime.datetime.strptime(date_string, "%d/%m/%Y").date()
+    return formatted_date
+
+
+def load_projects(filename):
+    projects = []
+    with open(filename) as infile:
+        infile.readline()
+        for line in infile:
+            parts = line.strip().split("\t")
+            start_date = format_date(parts[1])
+            project = Project(parts[0], start_date, int(parts[2]), float(parts[3]), int(parts[4]))
+            projects.append(project)
+    return projects
+
+
+def save_projects(filename, projects):
+    with open(filename, "w") as outfile:
+        print("Name	Start Date	Priority	Cost Estimate	Completion Percentage", file=outfile)
+        for project in projects:
+            print(f"{project.name}\t{project.date}\t{project.priority}\t"
+                  f"{project.cost}\t{project.completion}\t", file=outfile)
+
+
+main()
```

Comment:
This is some good code which matches the sample output for the most part, would just check the sorting after adding a project so it fully matches but great job! Also docstrings but nbd 😆